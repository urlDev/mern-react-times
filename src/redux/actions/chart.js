import axios from "axios";

export const FETCH_FOREX_SUCCESS = "FETCH_FOREX_SUCCESS";
export const FETCH_CHART_ERROR = "FETCH_CHART_ERROR";
export const SET_LOADING = "SET_LOADING";
export const SET_LOADING_TRUE = "SET_LOADING_TRUE";
export const CHANGE_MARKET_TYPE = "CHANGE_MARKET_TYPE";
export const CHANGE_MARKET_NAME = "CHANGE_MARKET_NAME";
export const SET_MARKET_DETAIL = "SET_MARKET_DETAIL";
export const SET_CHART_DATA = "SET_CHART_DATA";
export const CLEAN_CHART_DATA = "CLEAN_CHART_DATA";
export const SET_CHART_TIME_FRAME = "SET_CHART_TIME_FRAME";
export const SET_RATING = "SET_RATING";
export const SET_HOME_CHART_DATA = "SET_HOME_CHART_DATA";
export const CLEAN_STATE = "CLEAN_STATE";
export const GET_SEARCH_RESULTS = "GET_SEARCH_RESULTS";
export const SET_SEARCH_MARKET_DETAIL = "SET_SEARCH_MARKET_DETAIL";
export const CLEAR_SEARCH_RESULTS = "CLEAR_SEARCH_RESULTS";
export const OPEN_SEARCH_MODAL = "OPEN_SEARCH_MODAL";
export const CLOSE_SEARCH_MODAL = "CLOSE_SEARCH_NODAL";

export const closeSearchModal = () => ({
  type: CLOSE_SEARCH_MODAL,
});

export const openSearchModal = () => ({
  type: OPEN_SEARCH_MODAL,
});

export const clearSearchResults = () => ({
  type: CLEAR_SEARCH_RESULTS,
});

export const setSearchMarketDetail = (details) => ({
  type: SET_SEARCH_MARKET_DETAIL,
  payload: details,
});

export const getSearchResults = (result) => ({
  type: GET_SEARCH_RESULTS,
  payload: result,
});

export const setLoading = () => ({
  type: SET_LOADING,
});

export const setLoadingTrue = () => ({
  type: SET_LOADING_TRUE,
});

export const cleanState = () => ({
  type: CLEAN_STATE,
});

export const setHomeChartData = (chart) => ({
  type: SET_HOME_CHART_DATA,
  payload: [chart],
});

export const setRating = (rating) => ({
  type: SET_RATING,
  payload: rating,
});

export const setChartTimeFrame = (time) => ({
  type: SET_CHART_TIME_FRAME,
  payload: time,
});

export const setChartData = (data) => ({
  type: SET_CHART_DATA,
  payload: data,
});

export const cleanChartData = () => ({
  type: CLEAN_CHART_DATA,
});

export const setMarketDetail = (market) => ({
  type: SET_MARKET_DETAIL,
  payload: market,
});

export const changeMarketType = (market) => ({
  type: CHANGE_MARKET_TYPE,
  payload: market,
});

export const changeMarketName = (market) => ({
  type: CHANGE_MARKET_NAME,
  payload: market,
});

export const fetchForexSuccess = (chart) => ({
  type: FETCH_FOREX_SUCCESS,
  payload: chart,
});

export const fetchChartError = (error) => ({
  type: FETCH_CHART_ERROR,
  payload: error,
});

export const fetchForex = (market) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/quote/${market}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    // FinancialModel API doesn't send error if it's about limit reach.
    // In that sense, they send an error message with 200 status. Weird.
    // Thats why I am catching that error here and dispatching it.
    if (data.length) {
      return data.length === 1
        ? [dispatch(setSearchMarketDetail(data))]
        : dispatch(fetchForexSuccess(data));
    } else {
      return dispatch(fetchChartError(data));
    }
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchChartData = (symbol, timeFrame) => async (dispatch) => {
  dispatch(cleanChartData());
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/historical-chart/${timeFrame}/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    return dispatch(setChartData(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchRating = (symbol) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/rating/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    return dispatch(setRating(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

/*
This deserves to be mentioned.
Few months ago, when I was making the first version of this app, I used array.map to fetch charts.
Because of the async nation and using array map method, response was coming irregular and I had to sort them.
Now, with using for of loop, I dont need to write any extra code because loop waits for one to finish to move on.
Plus, its so clean!
Yet another homerun for me! 😁😎
*/
export const fetchHomeChart = (symbols) => async (dispatch) => {
  const symbolsArray = symbols.split(",").sort();
  let charts = [];
  try {
    // dispatch(setLoadingTrue());
    dispatch(cleanState());
    for (let symbol of symbolsArray) {
      const response = await axios.get(
        `https://financialmodelingprep.com/api/v3/historical-chart/1hour/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
      );
      const data = response.data;
      charts.push(data);
    }
    return charts.length
      ? dispatch(setHomeChartData(charts))
      : dispatch(fetchChartError(charts));
    // dispatch(setLoading());
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchSearch = (input) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/search?query=${input}&limit=10&apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = response.data;
    return dispatch(getSearchResults(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};
