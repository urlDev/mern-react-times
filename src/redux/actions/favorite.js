import axios from "axios";
import toaster from "toasted-notes";

import NotificationComponent from "../../components/notification-component/NotificationComponent";

const url = "https://urldev-mern-react-times-api.herokuapp.com";
// const url = `http://localhost:3000`;

export const GET_FAVORITE = "GET_FAVORITE";
export const ADD_FAVORITE = "ADD_FAVORITE";
export const DELETE_FAVORITE = "DELETE_FAVORITE";
export const FETCH_FAVORITE_ERROR = "FETCH_FAVORITE_ERROR";
export const CLEAN_FAVORITE_STATE = "CLEAN_FAVORITE_STATE";

export const getFavorite = (stock) => ({
  type: GET_FAVORITE,
  payload: stock,
});

export const addFavorite = (stock) => ({
  type: ADD_FAVORITE,
  payload: stock,
});

export const deleteFavorite = (stock) => ({
  type: DELETE_FAVORITE,
  payload: stock,
});

export const fetchFavoriteError = (error) => ({
  type: FETCH_FAVORITE_ERROR,
  payload: error,
});

export const cleanFavoriteState = () => ({
  type: CLEAN_FAVORITE_STATE,
});

export const fetchGetFavorites = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem("token"));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    const response = await axios.get(`${url}/`, config);
    const data = await response.data;
    return dispatch(getFavorite(data));
  } catch (error) {
    return dispatch(fetchFavoriteError(error));
  }
};

export const fetchAddFavorites = (stock) => async (dispatch) => {
  const favoriteStock = {
    symbol: stock,
  };

  const token = JSON.parse(localStorage.getItem("token"));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const response = await axios.post(`${url}/`, favoriteStock, config);
    const data = await response.data;
    return [
      dispatch(addFavorite(data.symbol[0])),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`You added ${data.symbol[0].name
              .split("^")
              .join("")} to your favorites successfully.`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(fetchFavoriteError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            success={false}
            text={`We just broke something, try again please or refresh.`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchDeleteFavorite = (id) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem("token"));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    const response = await axios.delete(`${url}/${id}`, config);
    const data = await response.data;
    return [
      dispatch(deleteFavorite(data)),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`You deleted ${data.symbol[0].name
              .split("^")
              .join("")} from your favorites.`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(fetchFavoriteError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            success={false}
            text={`We just broke something, try again please or refresh.`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};
