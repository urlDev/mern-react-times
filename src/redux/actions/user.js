import axios from 'axios';
import toaster from 'toasted-notes';
import { push } from 'connected-react-router';

import NotificationComponent from '../../components/notification-component/NotificationComponent';

const url = process.env.REACT_APP_API_URL;

export const REGISTER_USER = 'REGISTER_USER';
export const LOGOUT_USER = 'LOGOUT_USER';
export const LOGIN_USER = 'LOGIN_USER';
export const DELETE_USER = 'DELETE_USER';
export const UPDATE_USER = 'UPDATE_USER';
export const GET_USER = 'GET_USER';
export const SET_TOKEN = 'SET_TOKEN';
export const USER_FETCH_ERROR = 'USER_FETCH_ERROR';
export const USER_LOADING = 'USER_LOADING';
export const USER_LOADING_END = 'USER_LOADING_END';
export const USER_MODAL_OPEN = 'USER_MODAL_OPEN';
export const USER_MODAL_CLOSE = 'USER_MODAL_CLOSE';
export const DELETE_MODAL_OPEN = 'DELETE_MODAL_OPEN';
export const DELETE_MODAL_CLOSE = 'DELETE_MODAL_CLOSE';
export const UPLOAD_AVATAR = 'UPLOAD_AVATAR';

export const registerUser = (user) => ({
  type: REGISTER_USER,
  payload: user,
});

export const logOutUser = () => ({
  type: LOGOUT_USER,
});

export const loginUser = (user) => ({
  type: LOGIN_USER,
  payload: user,
});

export const deleteUser = () => ({
  type: DELETE_USER,
});

export const updateUser = (user) => ({
  type: UPDATE_USER,
  payload: user,
});

export const getUser = (user) => ({
  type: GET_USER,
  payload: user,
});

export const setToken = (token) => ({
  type: SET_TOKEN,
  payload: token,
});

export const userFetchError = (error) => ({
  type: USER_FETCH_ERROR,
  payload: error,
});

export const userLoading = () => ({
  type: USER_LOADING,
});

export const userLoadingEnd = () => ({
  type: USER_LOADING_END,
});

export const userModalOpen = () => ({
  type: USER_MODAL_OPEN,
});

export const userModalClose = () => ({
  type: USER_MODAL_CLOSE,
});

export const deleteModalOpen = () => ({
  type: DELETE_MODAL_OPEN,
});

export const deleteModalClose = () => ({
  type: DELETE_MODAL_CLOSE,
});

export const uploadAvatar = (images) => ({
  type: UPLOAD_AVATAR,
  payload: images,
});

export const fetchRegisterUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/register`, user);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchLogoutUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  const path = window.location.pathname;

  try {
    await axios.post(`${url}/profile/logout`, null, config);

    return [
      path.includes('profile') && dispatch(push('/home')),
      dispatch(logOutUser()),
      toaster.notify(
        () => <NotificationComponent text={'Buh-Bye!'} success={true} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return dispatch(userFetchError(error));
  }
};

export const fetchLoginUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/login`, user);
    const data = await response.data;
    return [
      dispatch(loginUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/home')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

// To get the user,
// but because I added an or statement in user and token initial states,
// app will check if there is user and token in localStorage
// so there will be no need for this.
// Adding just in case.
export const fetchUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  dispatch(userLoading());
  try {
    const response = await axios.get(`${url}/profile`, config);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
    ];
  } catch (error) {
    return [dispatch(userFetchError(error)), dispatch(userLoadingEnd())];
  }
};

export const fetchUpdateUser = (user) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    const response = await axios.patch(`${url}/profile`, user, config);
    const data = await response.data;
    return [
      dispatch(updateUser(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Updated successfully!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchDeleteUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    await axios.delete(`${url}/profile`, config);
    return [
      localStorage.clear(),
      dispatch(push('/')),
      dispatch(deleteUser()),
      toaster.notify(
        () => <NotificationComponent success={true} text={`Buh-bye!`} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            success={false}
            text={`Something went wrong!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchUploadAvatar = (input) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const response = await axios.post(`${url}/profile/avatar`, input, config);
    const data = await response.data;
    return [
      dispatch(uploadAvatar(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Avatar is changed! Wow, that looks amazing! ;)'}
            success={true}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};
