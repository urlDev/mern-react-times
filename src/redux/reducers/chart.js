import {
    FETCH_FOREX_SUCCESS,
    FETCH_CHART_ERROR,
    CHANGE_MARKET_TYPE,
    SET_MARKET_DETAIL,
    SET_CHART_DATA,
    SET_CHART_TIME_FRAME,
    SET_RATING,
    SET_HOME_CHART_DATA,
} from '../actions/chart';

import chartSort from '../utils/chartSort';

const initialState = {
    loading: true,
    forex: [],
    error: null,
    // Indexes are first to load.
    marketType: '%5EGSPC,%5ERUA,%5EDJI,%5ENDX,%5EN225,%5EFTSE',
    marketDetail: {},
    chartData: [],
    chartTimeFrame: '1hour',
    rating: [],
    homeChartData: [],
};

const chartReducer = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_FOREX_SUCCESS:
            return {
                ...state,
                forex: action.payload,
                loading: false,
            };
        case FETCH_CHART_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case CHANGE_MARKET_TYPE:
            return {
                ...state,
                marketType: action.payload,
            };
        case SET_MARKET_DETAIL:
            return {
                ...state,
                marketDetail: action.payload.data,
            };
        case SET_CHART_DATA:
            return {
                ...state,
                chartData: action.payload,
                loading: false,
            };
        case SET_CHART_TIME_FRAME:
            return {
                ...state,
                chartTimeFrame: action.payload,
            };
        case SET_RATING:
            return {
                ...state,
                rating: action.payload,
            };
        case SET_HOME_CHART_DATA:
            return {
                ...state,
                homeChartData: [...state.homeChartData, action.payload],
                loading: false,
            };
        default:
            return state;
    }
};

export default chartReducer;