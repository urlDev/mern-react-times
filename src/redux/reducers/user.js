import {
    REGISTER_USER,
    SET_TOKEN,
    USER_FETCH_ERROR,
    USER_LOADING,
    USER_LOADING_END,
} from '../actions/user';

const initialState = {
    loadingUser: false,
    errorUser: null,
    user: JSON.parse(localStorage.getItem('user')) || {},
    token: JSON.parse(localStorage.getItem('token')) || {},
};

const userReducer = (state = initialState, action) => {
    switch (action.payload) {
        case REGISTER_USER:
            return {
                ...state,
                user: action.payload,
            };
        case SET_TOKEN:
            return {
                ...state,
                token: action.payload,
            };
        case USER_FETCH_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case USER_LOADING:
            return {
                ...state,
                loading: true,
            };
        case USER_LOADING_END:
            return {
                ...state,
                loading: false,
            };
        default:
            return state;
    }
};

export default userReducer;