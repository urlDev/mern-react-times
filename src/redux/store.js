import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));