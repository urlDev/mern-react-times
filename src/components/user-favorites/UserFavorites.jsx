import React from "react";
import { useSelector, useDispatch } from "react-redux";

import {
  fetchGetFavorites,
  fetchDeleteFavorite,
} from "../../redux/actions/favorite";
import { setMarketDetail, fetchRating } from "../../redux/actions/chart";

import Nav from "../nav/Nav";
import StoryTopicHeaders from "../story-topic-headers/StoryTopicHeaders";

import LogoDarkSrc from "../../assets/logo.svg";
import AddedSrc from "../../assets/bookmark.svg";

import {
  StoryTopicContainer,
  FavoriteButton,
} from "../story-topic/StoryTopic.styles.js";
import {
  UserFavoriteContainer,
  FavoriteCards,
  StockPercentage,
} from "./UserFavorites.styles";

const UserFavorites = () => {
  const { favorites } = useSelector((favorite) => favorite.favorite);
  const { user } = useSelector((user) => user.user);
  const dispatch = useDispatch();

  React.useEffect(() => {
    if (user.name) {
      dispatch(fetchGetFavorites());
    }
  }, [user.name, dispatch]);

  return (
    <>
      <Nav logo={LogoDarkSrc} borderBottom="1px solid lightgray" icon="1" />
      <StoryTopicHeaders />
      <div style={{ margin: "0 30px" }}>
        <StoryTopicContainer>
          <h1 style={{ marginTop: "25px" }}>Your Favorites</h1>
        </StoryTopicContainer>
        <UserFavoriteContainer>
          {favorites.length ? (
            favorites.map((stock) => {
              const { symbol, price, changesPercentage } = stock.symbol[0];

              return (
                <div
                  style={{
                    display: "flex",
                    border: "1px solid black",
                    padding: "10px 15px",
                  }}
                >
                  <FavoriteButton
                    style={{ margin: "auto 10px auto 0" }}
                    onClick={() => dispatch(fetchDeleteFavorite(stock._id))}
                  >
                    <img src={AddedSrc} alt="favorite added, bookmark svg" />
                  </FavoriteButton>
                  <FavoriteCards
                    to={`/details/${symbol && symbol.toLowerCase()}`}
                    percentage={stock.changesPercentage}
                    onClick={() => {
                      dispatch(setMarketDetail(stock.symbol[0]));
                      dispatch(fetchRating(symbol));
                    }}
                  >
                    <h1>{symbol.split("^").join("")}</h1>

                    <h1
                      style={{
                        fontWeight: "normal",
                        fontSize: "var(--size-sub-menu)",
                      }}
                    >
                      $ {price.toFixed(2)}
                    </h1>
                    <StockPercentage percentage={changesPercentage}>
                      <h1>{changesPercentage} %</h1>
                    </StockPercentage>
                  </FavoriteCards>
                </div>
              );
            })
          ) : (
            <h1>No favorites added.</h1>
          )}
        </UserFavoriteContainer>
      </div>
    </> //
  );
};

export default UserFavorites;
