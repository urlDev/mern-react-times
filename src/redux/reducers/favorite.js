import * as FavoriteActions from "redux/actions/favorite";

const initialState = {
  errorFavorite: null,
  favorites: [],
};

const favoriteReducer = (state = initialState, action) => {
  switch (action.type) {
    case FavoriteActions.ADD_FAVORITE:
      return {
        ...state,
        favorites: [...state.favorites, action.payload],
      };
    case FavoriteActions.DELETE_FAVORITE:
      return {
        ...state,
        favorites: state.favorites.filter(
          (favorite) => favorite._id !== action.payload._id
        ),
      };
    case FavoriteActions.GET_FAVORITE:
      return {
        ...state,
        favorites: action.payload,
      };
    case FavoriteActions.FETCH_FAVORITE_ERROR:
      return {
        ...state,
        errorFavorite: action.payload,
      };
    case FavoriteActions.CLEAN_FAVORITE_STATE:
      return {
        ...state,
        favorites: [],
      };
    default:
      return state;
  }
};

export default favoriteReducer;
