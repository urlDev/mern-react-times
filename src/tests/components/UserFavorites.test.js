import React from "react";
import { shallow } from "enzyme";
import moxios from "moxios";
import renderer from "react-test-renderer";
import "jest-styled-components";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import { rating } from "tests/fixtures/chart";
import {
  fetchRating,
  SET_MARKET_DETAIL,
  SET_RATING,
} from "redux/actions/chart";
import { DELETE_FAVORITE, fetchDeleteFavorite } from "redux/actions/favorite";
import { user } from "tests/fixtures/user";
import { favorite, favorites } from "tests/fixtures/favorite";

import UserFavorites from "components/user-favorites/UserFavorites";
import { StockPercentage } from "components/user-favorites/UserFavorites.styles";

describe("Testing UserFavorites component", () => {
  let wrapper;
  let store;

  beforeEach(() => {
    jest.spyOn(React, "useEffect").mockImplementation((f) => f());

    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  describe("Testing component with favorites", () => {
    beforeEach(() => {
      store = mockStore({
        favorites,
        user,
      });

      wrapper = shallow(<UserFavorites store={store} />);
    });

    test("Should render the component with favorites", () => {
      expect(wrapper).toHaveLength(1);
      expect(wrapper.find("h1").contains("No favorites added.")).toBeFalsy();
    });
    test("Should match the snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });

    test("Should dispatch action on link click", async () => {
      wrapper.find("Styled(Link)").last().simulate("click");

      moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith({
          status: 200,
          response: rating,
        });
      });

      await store.dispatch(fetchRating("AMZN"));

      const actions = store.getActions();
      // onClick dispatches two actions
      // and returns markets details (favorite symbol)
      // and rating.
      expect(actions).toEqual([
        {
          type: SET_MARKET_DETAIL,
          payload: favorites[0].symbol[0],
        },
        {
          type: SET_RATING,
          payload: rating,
        },
      ]);
    });
    test("Should delete stock from favorites", async () => {
      wrapper.find("div").last().childAt(0).simulate("click");

      moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith({
          status: 200,
          response: favorite,
        });
      });

      await store.dispatch(fetchDeleteFavorite(favorite._id));

      const actions = store.getActions();
      expect(actions).toEqual([
        {
          type: DELETE_FAVORITE,
          payload: favorite,
        },
      ]);
    });
  });

  describe("Testing component without favorites", () => {
    test("Should render component without favorite stocks", () => {
      store = mockStore({
        user,
        favorites: [],
      });

      wrapper = shallow(<UserFavorites store={store} />);

      expect(wrapper).toHaveLength(1);
      expect(wrapper.find("h1").contains("No favorites added.")).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });
  test("Should apply styles to StockPercentage according to default", () => {
    wrapper = renderer.create(<StockPercentage percentage="-1" />).toJSON();
    // Default values in styles
    expect(wrapper).toHaveStyleRule("color", "var(--red)");
  });
  test("Should apply styles to StockPercentage according to passed props", () => {
    wrapper = renderer.create(<StockPercentage percentage="2" />).toJSON();
    // Passed props to component
    expect(wrapper).toHaveStyleRule("color", "var(--green)");
  });
});
