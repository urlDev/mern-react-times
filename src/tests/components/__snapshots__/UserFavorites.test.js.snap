// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UserFavorites component Testing component with favorites Should match the snapshot 1`] = `
<Fragment>
  <ResponsiveNav />
  <div
    style={
      Object {
        "margin": "0 15px",
      }
    }
  >
    <styled.div>
      <h1
        style={
          Object {
            "marginTop": "25px",
          }
        }
      >
        Your Favorites
      </h1>
    </styled.div>
    <styled.div>
      <div
        key="AMZN"
        style={
          Object {
            "border": "1px solid black",
            "display": "flex",
            "padding": "10px 15px",
          }
        }
      >
        <styled.button
          onClick={[Function]}
          style={
            Object {
              "margin": "auto 10px auto 0",
            }
          }
        >
          <img
            alt="favorite added, bookmark svg"
            src="bookmark.svg"
          />
        </styled.button>
        <Styled(Link)
          onClick={[Function]}
          to="/details/amzn"
        >
          <h1>
            AMZN
          </h1>
          <h1
            style={
              Object {
                "fontSize": "var(--size-sub-menu)",
                "fontWeight": "normal",
              }
            }
          >
            $ 
            3292.22
          </h1>
          <styled.div
            percentage={3.77}
          >
            <h1>
              3.77
               %
            </h1>
          </styled.div>
        </Styled(Link)>
      </div>
      <div
        key="AMZN"
        style={
          Object {
            "border": "1px solid black",
            "display": "flex",
            "padding": "10px 15px",
          }
        }
      >
        <styled.button
          onClick={[Function]}
          style={
            Object {
              "margin": "auto 10px auto 0",
            }
          }
        >
          <img
            alt="favorite added, bookmark svg"
            src="bookmark.svg"
          />
        </styled.button>
        <Styled(Link)
          onClick={[Function]}
          to="/details/amzn"
        >
          <h1>
            AMZN
          </h1>
          <h1
            style={
              Object {
                "fontSize": "var(--size-sub-menu)",
                "fontWeight": "normal",
              }
            }
          >
            $ 
            3292.22
          </h1>
          <styled.div
            percentage={3.77}
          >
            <h1>
              3.77
               %
            </h1>
          </styled.div>
        </Styled(Link)>
      </div>
    </styled.div>
  </div>
</Fragment>
`;

exports[`Testing UserFavorites component Testing component without favorites Should render component without favorite stocks 1`] = `
<Fragment>
  <ResponsiveNav />
  <div
    style={
      Object {
        "margin": "0 15px",
      }
    }
  >
    <styled.div>
      <h1
        style={
          Object {
            "marginTop": "25px",
          }
        }
      >
        Your Favorites
      </h1>
    </styled.div>
    <styled.div>
      <h1>
        No favorites added.
      </h1>
    </styled.div>
  </div>
</Fragment>
`;
